File: src/differential_equations_systems_of_odes_medium.py
"""
differential_equations - systems_of_odes (medium)
Generated: 2026-02-22T05:32:38.272872
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 4)
    
    if problem_type == 1:
        # Type 1: 2x2 linear system with constant coefficients (1300-1400)
        # Backward engineer: pick solution first
        x_sol = nonzero(-5, 5)
        y_sol = nonzero(-5, 5)
        
        # Create coefficients
        a11, a12 = nonzero(-4, 4), nonzero(-4, 4)
        a21, a22 = nonzero(-4, 4), nonzero(-4, 4)
        
        # Ensure system is not degenerate
        while a11 * a22 == a12 * a21:
            a22 = nonzero(-4, 4)
        
        # Compute right-hand sides
        b1 = a11 * x_sol + a12 * y_sol
        b2 = a21 * x_sol + a22 * y_sol
        
        # Format the system
        eq1_str = f"{latex(a11)}x + {latex(a12)}y = {latex(b1)}"
        eq2_str = f"{latex(a21)}x + {latex(a22)}y = {latex(b2)}"
        
        question = f"Solve the system of differential equations (find equilibrium point):<br>$\\frac{{dx}}{{dt}} = {eq1_str}$<br>$\\frac{{dy}}{{dt}} = {eq2_str}$"
        
        ans = fmt_tuple([x_sol, y_sol])
        
        solution = steps(
            f"Set $\\frac{{dx}}{{dt}} = 0$ and $\\frac{{dy}}{{dt}} = 0$ to find equilibrium",
            f"This gives us: ${eq1_str}$ and ${eq2_str}$",
            f"Using substitution or elimination methods",
            f"Solution: $(x, y) = ({x_sol}, {y_sol})$"
        )
        
        return problem(
            question=question,
            answer=ans,
            difficulty=(1300, 1400),
            topic="differential_equations/systems_of_odes",
            solution=solution,
            answer_type="tuple"
        )
    
    elif problem_type == 2:
        # Type 2: Solve dx/dt = ax, dy/dt = by (uncoupled system) (1300-1400)
        a_val = nonzero(-3, 3)
        b_val = nonzero(-3, 3)
        
        x0 = nonzero(-5, 5)
        y0 = nonzero(-5, 5)
        
        # Solutions are x(t) = x0*e^(a*t), y(t) = y0*e^(b*t)
        x_sol = x0 * exp(a_val * t)
        y_sol = y0 * exp(b_val * t)
        
        question = f"Solve the system with initial conditions $x(0) = {x0}$, $y(0) = {y0}$:<br>$\\frac{{dx}}{{dt}} = {latex(a_val)}x$<br>$\\frac{{dy}}{{dt}} = {latex(b_val)}y$"
        
        ans = fmt_tuple([x_sol, y_sol])
        
        solution = steps(
            f"These are uncoupled first-order linear ODEs",
            f"For $\\frac{{dx}}{{dt}} = {latex(a_val)}x$ with $x(0) = {x0}$: $x(t) = {latex(x_sol)}$",
            f"For $\\frac{{dy}}{{dt}} = {latex(b_val)}y$ with $y(0) = {y0}$: $y(t) = {latex(y_sol)}$",
            f"Solution: $(x(t), y(t)) = ({latex(x_sol)}, {latex(y_sol)})$"
        )
        
        return problem(
            question=question,
            answer=ans,
            difficulty=(1300, 1400),
            topic="differential_equations/systems_of_odes",
            solution=solution,
            answer_type="tuple"
        )
    
    elif problem_type == 3:
        # Type 3: Find eigenvalues of 2x2 system matrix (1400-1500)
        # Backward engineer: pick eigenvalues first
        lambda1 = nonzero(-4, 4)
        lambda2 = nonzero(-4, 4)
        
        # Create a matrix with these eigenvalues (diagonal or simple form)
        if randint(0, 1) == 0:
            # Diagonal matrix
            A = Matrix([[lambda1, 0], [0, lambda2]])
        else:
            # Upper triangular
            off_diag = nonzero(-3, 3)
            A = Matrix([[lambda1, off_diag], [0, lambda2]])
        
        question = f"Find the eigenvalues of the coefficient matrix for the system $\\frac{{d\\mathbf{{x}}}}{{dt}} = A\\mathbf{{x}}$ where $A = {latex(A)}$"
        
        ans = fmt_set([lambda1, lambda2])
        
        solution = steps(
            f"Find eigenvalues by solving $\\det(A - \\lambda I) = 0$",
            f"For $A = {latex(A)}$",
            f"Characteristic equation: $(\\lambda - {lambda1})(\\lambda - {lambda2}) = 0$",
            f"Eigenvalues: $\\lambda = {{{lambda1}, {lambda2}}}$"
        )
        
        return problem(
            question=question,
            answer=ans,
            difficulty=(1400, 1500),
            topic="differential_equations/systems_of_odes",
            solution=solution,
            answer_type="set"
        )
    
    else:
        # Type 4: Coupled 2x2 system - solve for one variable (1500-1600)
        # dx/dt = ax + by, dy/dt = cx + dy
        # Pick simpler values
        a_val = choice([-2, -1, 1, 2])
        b_val = choice([-2, -1, 1, 2])
        c_val = choice([-2, -1, 1, 2])
        d_val = choice([-2, -1, 1, 2])
        
        # Ensure trace^2 >= 4*det for real eigenvalues
        trace = a_val + d_val
        det = a_val * d_val - b_val * c_val
        
        if trace**2 < 4*det:
            d_val = -a_val
        
        A = Matrix([[a_val, b_val], [c_val, d_val]])
        eigenvals = list(A.eigenvals().keys())
        
        question = f"For the system $\\frac{{dx}}{{dt}} = {latex(a_val)}x + {latex(b_val)}y$, $\\frac{{dy}}{{dt}} = {latex(c_val)}x + {latex(d_val)}y$, classify the equilibrium at the origin based on eigenvalues."
        
        # Determine stability
        real_parts = [complex(e.evalf()).real for e in eigenvals]
        
        if all(r < 0 for r in real_parts):
            ans = "stable"
        elif all(r > 0 for r in real_parts):
            ans = "unstable"
        else:
            ans = "saddle"
        
        solution = steps(
            f"Find eigenvalues of $A = {latex(A)}$",
            f"Eigenvalues: ${latex(eigenvals[0])}, {latex(eigenvals[1])}$",
            f"Based on signs of real parts, the origin is {ans}"
        )
        
        return problem(
            question=question,
            answer=ans,
            difficulty=(1500, 1600),
            topic="differential_equations/systems_of_odes",
            solution=solution,
            answer_type="expression"
        )

emit(generate())

File: src/differential_equations_systems_of_odes_medium.jl
# differential_equations - systems_of_odes (medium)
# Generated: 2026-03-08T20:57:18.363763

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x y t begin
    set_topic!("differential_equations/systems_of_odes")
    
    problem_type = choice([:decoupled_system, :coupled_linear, :eigenvalue_system, :initial_value])
    
    if problem_type == :decoupled_system
        # System where equations are independent
        a1 = nonzero(-5, 5)
        a2 = nonzero(-5, 5)
        b1 = randint(-8, 8)
        b2 = randint(-8, 8)
        
        question_text = "Solve the system of differential equations:<br>" *
                       "\$\\frac{dx}{dt} = $(a1)x + $(b1)\$<br>" *
                       "\$\\frac{dy}{dt} = $(a2)y + $(b2)\$"
        
        if b1 == 0
            sol_x = "x(t) = C_1 e^{$(a1)t}"
        else
            sol_x = "x(t) = C_1 e^{$(a1)t} - $(b1//a1)"
        end
        
        if b2 == 0
            sol_y = "y(t) = C_2 e^{$(a2)t}"
        else
            sol_y = "y(t) = C_2 e^{$(a2)t} - $(b2//a2)"
        end
        
        answer_text = sol_x * ", " * sol_y
        
        solution_text = steps(
            "System is decoupled - solve each equation independently",
            sol("First equation", "\\frac{dx}{dt} = $(a1)x + $(b1)"),
            b1 == 0 ? "Separable: \$x(t) = C_1 e^{$(a1)t}\$" : "Linear ODE: \$x(t) = C_1 e^{$(a1)t} - \\frac{$(b1)}{$(a1)}\$",
            sol("Second equation", "\\frac{dy}{dt} = $(a2)y + $(b2)"),
            b2 == 0 ? "Separable: \$y(t) = C_2 e^{$(a2)t}\$" : "Linear ODE: \$y(t) = C_2 e^{$(a2)t} - \\frac{$(b2)}{$(a2)}\$",
            sol("General solution", answer_text)
        )
        
        problem(
            question=question_text,
            answer=answer_text,
            difficulty=(1200, 1400),
            solution=solution_text,
            time=180,
            answer_type="expression"
        )
        
    elseif problem_type == :coupled_linear
        # 2x2 coupled linear system with triangular structure
        a = nonzero(-4, 4)
        b = nonzero(-6, 6)
        c = nonzero(-4, 4)
        
        question_text = "Solve the system:<br>" *
                       "\$\\frac{dx}{dt} = $(a)x + $(b)y\$<br>" *
                       "\$\\frac{dy}{dt} = $(c)y\$"
        
        sol_y = "y(t) = C_2 e^{$(c)t}"
        
        if b == 0
            sol_x = "x(t) = C_1 e^{$(a)t}"
        else
            if a == c
                sol_x = "x(t) = C_1 e^{$(a)t} + $(b) C_2 t e^{$(a)t}"
            else
                coef = b // (c - a)
                sol_x = "x(t) = C_1 e^{$(a)t} + $(coef) C_2 e^{$(c)t}"
            end
        end
        
        answer_text = sol_x * ", " * sol_y
        
        solution_text = steps(
            "Start with the second equation (decoupled)",
            sol("\\frac{dy}{dt} = $(c)y", "y(t) = C_2 e^{$(c)t}"),
            "Substitute into first equation",
            sol("\\frac{dx}{dt} = $(a)x + $(b)C_2 e^{$(c)t}", ""),
            "Solve linear first-order ODE using integrating factor",
            sol("Solution", answer_text)
        )
        
        problem(
            question=question_text,
            answer=answer_text,
            difficulty=(1400, 1600),
            solution=solution_text,
            time=240,
            answer_type="expression"
        )
        
    elseif problem_type == :eigenvalue_system
        # Simple eigenvalue problem with real distinct eigenvalues
        lambda1 = nonzero(-6, 6)
        lambda2 = nonzero(-6, 6)
        while lambda2 == lambda1
            lambda2 = nonzero(-6, 6)
        end
        
        # Construct matrix with eigenvalues lambda1, lambda2
        # Use diagonal or simple form
        if choice([true, false])
            a11, a22 = lambda1, lambda2
            a12, a21 = 0, 0
            v1_text = "\\begin{pmatrix} 1 \\\\ 0 \\end{pmatrix}"
            v2_text = "\\begin{pmatrix} 0 \\\\ 1 \\end{pmatrix}"
        else
            # Create 2x2 with trace = lambda1 + lambda2, det = lambda1*lambda2
            trace_val = lambda1 + lambda2
            det_val = lambda1 * lambda2
            a11 = randint(-3, 3)
            a22 = trace_val - a11
            a12 = nonzero(-4, 4)
            a21 = (det_val - a11*a22) // a12
            v1_text = "\\begin{pmatrix} $(a12) \\\\ $(lambda1 - a11) \\end{pmatrix}"
            v2_text = "\\begin{pmatrix} $(a12) \\\\ $(lambda2 - a11) \\end{pmatrix}"
        end
        
        question_text = "Find the eigenvalues of the system:<br>" *
                       "\$\\frac{d\\mathbf{x}}{dt} = \\begin{pmatrix} $(a11) & $(a12) \\\\ $(a21) & $(a22) \\end{pmatrix} \\mathbf{x}\$"
        
        answer_text = fmt_set(Set([lambda1, lambda2]))
        
        solution_text = steps(
            sol("Matrix", "A = \\begin{pmatrix} $(a11) & $(a12) \\\\ $(a21) & $(a22) \\end{pmatrix}"),
            "Find eigenvalues by solving \$\\det(A - \\lambda I) = 0\$",
            sol("Characteristic equation", "($(a11) - \\lambda)($(a22) - \\lambda) - ($(a12))($(a21)) = 0"),
            "Expand and solve quadratic equation",
            sol("Eigenvalues", "\\lambda = $(lambda1), $(lambda2)")
        )
        
        problem(
            question=question_text,
            answer=answer_text,
            difficulty=(1400, 1700),
            solution=solution_text,
            time=200,
            answer_type="set"
        )
        
    else  # initial_value
        # System with initial conditions
        a = nonzero(-5, 5)
        b = nonzero(-5, 5)
        x0 = randint(-10, 10)
        y0 = randint(-10, 10)
        
        question_text = "Solve the initial value problem:<br>" *
                       "\$\\frac{dx}{dt} = $(a)x\$, \$x(0) = $(x0)\$<br>" *
                       "\$\\frac{dy}{dt} = $(b)y\$, \$y(0) = $(y0)\$"
        
        sol_x = "x(t) = $(x0) e^{$(a)t}"
        sol_y = "y(t) = $(y0) e^{$(b)t}"
        answer_text = sol_x * ", " * sol_y
        
        solution_text = steps(
            "Solve each decoupled equation separately",
            sol("For x(t)", "\\frac{dx}{dt} = $(a)x \\implies x(t) = C_1 e^{$(a)t}"),
            "Apply initial condition \$x(0) = $(x0)\$: \$C_1 = $(x0)\$",
            sol("For y(t)", "\\frac{dy}{dt} = $(b)y \\implies y(t) = C_2 e^{$(b)t}"),
            "Apply initial condition \$y(0) = $(y0)\$: \$C_2 = $(y0)\$",
            sol("Particular solution", answer_text)
        )
        
        problem(
            question=question_text,
            answer=answer_text,
            difficulty=(1300, 1500),
            solution=solution_text,
            time=150,
            answer_type="expression"
        )
    end
end